\documentclass[10pt,oneside]{article}
\usepackage[sc]{mathpazo}
\usepackage[T1]{fontenc}
\usepackage{geometry}
\geometry{verbose,tmargin=2.5cm,bmargin=2.5cm,lmargin=2.5cm,rmargin=2.5cm}
\setcounter{secnumdepth}{2}
\setcounter{tocdepth}{2}
\usepackage{url}
\usepackage[unicode=true,pdfusetitle,
 bookmarks=true,bookmarksnumbered=true,bookmarksopen=true,bookmarksopenlevel=2,
 breaklinks=false,pdfborder={0 0 1},backref=false,colorlinks=false]
 {hyperref}
\hypersetup{
 pdfstartview={XYZ null null 1}}
\usepackage{breakurl}
\usepackage{epsfig}
\usepackage{subfigure}
\renewcommand{\familydefault}{\sfdefault} 
%\usepackage{biolinum}
\usepackage{Sweave}
\begin{document}
\input{SeqGL-concordance}



% <<setup, include=FALSE, cache=FALSE>>=
% library(knitr)
% # set global chunk options
% opts_chunk$set(fig.path='figure/minimal-', fig.align='center', fig.show='hold')
% options(replace.assign=TRUE,width=90)
% @


\title{\bf{Identification of TF binding profiles from ChIP-seq and Dnase-seq using SeqGL}}
\author{Manu Setty$^1$, Christina Leslie\\[1em]Computational Biology Program, MSKCC,\\ New York\\
\texttt{$^1$manu@cbio.mskcc.org}}



\maketitle

\begin{abstract}
SeqGL is a new group lasso-based algorithm to extract multiple transcription factor (TF) binding signals from ChIP- and DNase-seq profiles. Benchmarked on over 100 ChIP-seq experiments, SeqGL outperformed traditional motif discovery tools in discriminative accuracy and cofactor detection. SeqGL successfully scales to DNase-seq data, identifying a large multiplicity of TF signals confirmed by ChIP, and can be used with multitask training to learn genomic-context and cell-type specific TF signals.
\end{abstract}


\tableofcontents

\section{Installation}
The R package can be downloaded from \url{https://bitbucket.org/leslielab/seqgl}. 
\subsection {Dependencies}

\begin{itemize}
\item The following bioconductor packages: Biostrings, GenomicRanges, BSgenome, WGCNA, fastcluster, gtools, sfsmisc, kernlab. 
These packages can be installed using the command 
\begin{Schunk}
\begin{Sinput}
> source("http://bioconductor.org/biocLite.R")
> biocLite(c("Biostrings", "GenomicRanges", "BSgenome", "WGCNA", "fastcluster", "gtools", "sfsmisc", "kernlab"))
\end{Sinput}
\end{Schunk}

\item ChIPKernels package for wildcard kernel. This package can be downloaded from \url{https://bitbucket.org/leslielab/chipkernels}.

\item spams toolbox for running group lasso. Download and install the R package from \url{http://spams-devel.gforge.inria.fr/downloads.html}. Please refer to \url{http://www.thecoatlessprofessor.com/programming/rcpp-rcpparmadillo-and-os-x-mavericks-lgfortran-and-lquadmath-error} for issues with installing spams on OS X Mavericks.

\item HOMER motif finding tool for associating groups with motifs. \url{http://homer.salk.edu/homer/}.

\item BSgenome package for the organisms of your choice from Bioconductor. Example if the peaks are from hg19, install BSgenome.Hsapiens.UCSC.hg19 from \url{http://bioconductor.org/packages/release/data/annotation/html/BSgenome.Hsapiens.UCSC.hg19.html}. The example detailed in this vignette assumes this package has been installed. Install the HOMER genome as well for the organism of your choice using \texttt{perl /path-to-homer/configureHomer.pl -install hg19}.

\end{itemize}

After all the dependencies are installed, SeqGL can be installed from source using \texttt{R CMD INSTALL <path to package>}.

\section{Inputs}
Chip-seq or DNase-seq peaks are the inputs to SeqGL. The peaks should be provided in bed format and should contain the following columns.
\begin{itemize}
\item chrom: Chromosome 
\item start: Genomic start
\item end: Genomic end
\item strand: Strand
\item score: Score to rank the peaks. Can be ${-log (p-value)}$.
\item summit: Summit position in the peak
\item name: Unique identifier for each peak
\end{itemize}

An example peaks file can be found in the package. This bed file contains the top peaks for IRF4 ChIP-seq in GM12878, a lymphoblastoid cell line.
\begin{Schunk}
\begin{Sinput}
> peaks.file <- system.file( "extdata/gm12878_top_irf4_peaks.bed", package="SeqGL" )
> peaks <- read.table (peaks.file, header=TRUE)
> head (peaks)
\end{Sinput}
\begin{Soutput}
  chrom chromStart  chromEnd strand   score summit  name
1  chr1   37754504  37754786      * 2089.17    149 Peak1
2  chr1  181068642 181068903      *  916.04    136 Peak2
3  chr1   45965797  45966083      * 1573.60    162 Peak3
4  chr1  163291458 163291723      * 1876.17    107 Peak4
5  chr1  235114142 235114503      * 1753.84    127 Peak5
6  chr1   28526489  28526779      * 1525.03    160 Peak6
\end{Soutput}
\end{Schunk}


\section{SeqGL wrapper}
SeqGL has a wrapper function \texttt{run.seqGL} which takes the peaks, organism as inputs and run through the complete pipeline.
\begin{Schunk}
\begin{Sinput}
> peaks.file <- system.file( "extdata/gm12878_top_irf4_peaks.bed", package="SeqGL" )
> run.seqGL (peaks.file, out.dir="seqGL.Test/", data.type="ChIP", org="hg19")
\end{Sinput}
\end{Schunk}

The results of the will be present in the \texttt{seqGL.Test} folder and contains the following objects. 
The test performance is shown in the file \texttt{seqGL.Test/test\_auc.pdf} which shows the ROC plot. See Figure \ref{fig:fig1} for an example.

\begin{figure}[h!]
\begin{center}
\includegraphics[width=.75\textwidth]{figures/auROC}
\end{center}
\caption{\textbf{Test auROC for IRF4 peaks}}\label{fig:fig1}
\end{figure}

The motifs associated with each group can be found in \texttt{seqGL.Test/group\_motifs.html}. A screenshot is shown in Figure \ref{fig:fig2}.
\begin{figure}[h!]
\begin{center}
\includegraphics[width=.75\textwidth]{figures/motifs}
\end{center}
\caption{\textbf{Group scores and motifs for IRF peaks}}\label{fig:fig2}
\end{figure}

\pagebreak

The contents of the results folder:
\begin{itemize}
\item group\_motifs.html: Html file containing the group scores and motif associated with each group.
\item test\_auc.pdf: ROC plot showing the performance of the method. auROC varies from $0.5-1$ with being perfect classification.
\item group\_members: Folder containing peaks associated with each group.
\item group\_motifs: Folder containing all the HOMER results.
\item train\_test\_data.Rds R object containing the train and test data.
\item clustering\_results.Rds R object containing clustering results.
\item group\_lasso\_results.Rdata R object containing all the group lasso results.
\item seqgl_results.Rdata R object containing group scores, group membership and peak scores for all peaks.
\end{itemize}


\section{SeqGL details}
This section describes the different steps underlying SeqGL using the IRF4 peaks as an example. The following command will load the library.
\begin{Schunk}
\begin{Sinput}
> library (SeqGL)
\end{Sinput}
\begin{Soutput}
==========================================================================
*
*  Package WGCNA 1.46 loaded.
*
*    Important note: It appears that your system supports multi-threading,
*    but it is not enabled within WGCNA in R. 
*    To allow multi-threading within WGCNA with all available cores, use 
*
*          allowWGCNAThreads()
*
*    within R. Use disableWGCNAThreads() to disable threading if necessary.
*    Alternatively, set the following environment variable on your system:
*
*          ALLOW_WGCNA_THREADS=<number_of_processors>
*
*    for example 
*
*          ALLOW_WGCNA_THREADS=8
*
*    To set the environment variable in linux bash shell, type 
*
*           export ALLOW_WGCNA_THREADS=8
*
*     before running R. Other operating systems or shells will
*     have a similar command to achieve the same aim.
*
==========================================================================
\end{Soutput}
\end{Schunk}

\subsection{Getting data ready for SeqGL}

The first step is to normalize the spans of different peaks and ensure they are all of the same width. We recommed a span of $150$ bases around the peak summit for optimal performance.
\begin{Schunk}
\begin{Sinput}
> peaks.gr <- GRanges (peaks$chrom, IRanges (peaks$chromStart, peaks$chromEnd),
+ 	summit=peaks$summit, name=peaks$name)
> span <- 150
> pos.regions <- peaks.gr 
> start (pos.regions) <- start (pos.regions) + pos.regions$summit - 1
> end (pos.regions) <- start (pos.regions)
> pos.regions <- resize (pos.regions, span, fix='center')
\end{Sinput}
\end{Schunk}

Then create positive and negative regions. Negative regions are created by shifting the positive regions upstream.
\begin{Schunk}
\begin{Sinput}
> neg.regions <- shift (pos.regions, span*2)
\end{Sinput}
\end{Schunk}

After creating the regions or examples, we build the feature matriices for group lasso. The \texttt{build.features.kernels} function from \texttt{ChIPKernels} package is used for constructing the feature matrices. Wildcard string kernels are used for determining feature matrices. The \texttt{build.train.test.data} function splits the examples into train and test sets, determines sequences for all the examples and then builds the feature matrices. BSgenome package corresponding to the organims should be installed for determining sequences. Specifically, the peaks are for hg19 genome and \texttt{BSgenome.Hsapiens.UCSC.hg19} has to be installed in this example.

\begin{Schunk}
\begin{Sinput}
> res.dir <- '/tmp/SeqGLTest'; dir.create (res.dir)
> dictionary.file <- system.file( "extdata/wildcard_dict_kmer8_mismatches2_alpha5_consecutive_mis.Rdata", package="SeqGL" )
> train.test.data <- build.train.test.data (pos.regions, neg.regions, dictionary.file, org='hg19')
\end{Sinput}
\begin{Soutput}
[1] "Extract sequences for all training and test examples..."
[1] "Time for extracting sequences: 0.75 minutes"
[1] "Building features from dictionary..."
[1] "Position 1 of 143"
[1] "Position 21 of 143"
[1] "Position 41 of 143"
[1] "Position 61 of 143"
[1] "Position 81 of 143"
[1] "Position 101 of 143"
[1] "Position 121 of 143"
[1] "Position 141 of 143"
[1] "Time for determining features : 1.08"
[1] "Selecting top features..."
[1] "Time for selecting features: 0.17 minutes"
[1] "Package and return..."
[1] "Total time for constructing data: 2.02 minutes"
\end{Soutput}
\begin{Sinput}
> saveRDS (train.test.data, file=sprintf ("%s/train_test_data.Rds", res.dir))
\end{Sinput}
\end{Schunk}

\texttt{train.test.data} is a list containing all the training and test data
\begin{Schunk}
\begin{Sinput}
> show (labels (train.test.data))
\end{Sinput}
\begin{Soutput}
 [1] "train.features"  "test.features"   "train.inds"      "test.inds"      
 [5] "feature.inds"    "train.labels"    "test.labels"     "train.regions"  
 [9] "test.regions"    "dictionary.file" "train.seqs"      "test.seqs"      
\end{Soutput}
\end{Schunk}

\subsection{Identification of groups}

The groups are identified by hierarchical clustering of features. \texttt{run.clustering} function to used for hierarchical clustering.
\begin{Schunk}
\begin{Sinput}
> clustering.results <- run.clustering (train.test.data$train.features, no.groups=20)
\end{Sinput}
\begin{Soutput}
[1] "Running clustering..."
[1] "Time for running clustering: 1.63 minutes"
\end{Soutput}
\begin{Sinput}
> saveRDS (clustering.results, file=sprintf ("%s/clustering_results.Rds", res.dir))
\end{Sinput}
\end{Schunk}



\subsection{Group lasso}

The groups of kmers are used in a group lasso learning framework. We use the \texttt{spams} toolbox ro run group lasso. We first identify the optimal regularization parameters for group lasso and learn the model using these parameters. The functions \texttt{group.lasso.eval.parameters} and \texttt{run.group.lasso} are used for parameter evaulation and running group lasso respectively.
\begin{Schunk}
\begin{Sinput}
> lambdas=c(1e-2, 5e-3, 1e-3, 5e-4, 1e-4)
> param.eval <- group.lasso.eval.parameters (train.test.data$train.features, 
+ 	train.test.data$train.labels, train.test.data$test.features, 
+ 	train.test.data$test.labels, clustering.results$groups, 
+ 	lambdas=lambdas)
\end{Sinput}
\begin{Soutput}
[1] "Running crossvalidation ..."
[1] "i: 1 of 5, j: 1 of 5"
[1] "i: 1 of 5, j: 2 of 5"
[1] "i: 1 of 5, j: 3 of 5"
[1] "i: 1 of 5, j: 4 of 5"
[1] "i: 1 of 5, j: 5 of 5"
[1] "i: 2 of 5, j: 1 of 5"
[1] "i: 2 of 5, j: 2 of 5"
[1] "i: 2 of 5, j: 3 of 5"
[1] "i: 2 of 5, j: 4 of 5"
[1] "i: 2 of 5, j: 5 of 5"
[1] "i: 3 of 5, j: 1 of 5"
[1] "i: 3 of 5, j: 2 of 5"
[1] "i: 3 of 5, j: 3 of 5"
[1] "i: 3 of 5, j: 4 of 5"
[1] "i: 3 of 5, j: 5 of 5"
[1] "i: 4 of 5, j: 1 of 5"
[1] "i: 4 of 5, j: 2 of 5"
[1] "i: 4 of 5, j: 3 of 5"
[1] "i: 4 of 5, j: 4 of 5"
[1] "i: 4 of 5, j: 5 of 5"
[1] "i: 5 of 5, j: 1 of 5"
[1] "i: 5 of 5, j: 2 of 5"
[1] "i: 5 of 5, j: 3 of 5"
[1] "i: 5 of 5, j: 4 of 5"
[1] "i: 5 of 5, j: 5 of 5"
[1] "Time for running crossvalidation: 0.70 minutes"
\end{Soutput}
\begin{Sinput}
> saveRDS (param.eval, file=sprintf ("%s/param_eval.Rds", res.dir))
\end{Sinput}
\end{Schunk}

\begin{Schunk}
\begin{Sinput}
> inds <- which (param.eval$aucs.matrix == max (param.eval$aucs.matrix), arr.ind=TRUE)
> group.lasso.results <- run.group.lasso (train.test.data$train.features, train.test.data$train.labels,
+ 	train.test.data$test.features, train.test.data$test.labels, clustering.results$groups,
+ 	param.eval$lambdas[inds[1]], param.eval$lambdas[inds[2]])
\end{Sinput}
\begin{Soutput}
[1] "Running group lasso..."
[1] "Time for running group lasso: 0.03 minutes"
\end{Soutput}
\begin{Sinput}
> 
\end{Sinput}
\end{Schunk}


Group rankings and peaks associated with group can be determined using 
\begin{Schunk}
\begin{Sinput}
> peak.scores <- determine.peak.scores (train.test.data$train.features, 
+ 	train.test.data$train.labels, group.lasso.results$w, clustering.results$groups)
> group.scores <- determine.group.scores (train.test.data$train.features, 
+ 	train.test.data$train.labels, group.lasso.results$w, clustering.results$groups)
\end{Sinput}
\begin{Soutput}
[1] "Time for determining group scores: 0.00"
\end{Soutput}
\begin{Sinput}
> group.members <- determine.group.members  (train.test.data$train.labels, 
+ 	group.scores, peak.scores, 
+ 	test.classes=1)$group.members
\end{Sinput}
\begin{Soutput}
[1] "Time for determining group members: 0.00"
\end{Soutput}
\begin{Sinput}
> save (group.lasso.results, peak.scores, group.scores, group.members,
+ 	file=sprintf ("%s/group_lasso_results.Rdata", res.dir))
\end{Sinput}
\end{Schunk}

Finally, motifs can be generated using HOMER by invoking the function \texttt{group.motifs}
\begin{Schunk}
\begin{Sinput}
> group.motifs (res.dir, dictionary.file,
+ 	no.cores=1, org="hg19", test.classes=1)
\end{Sinput}
\end{Schunk}


\end{document}












